tests/core/test_hdaccount.py

```diff
@@ -0,0 +1,88 @@
+import pytest
+
+from eth_utils import (
+    ValidationError,
+)
+
+from eth_account import (
+    Account,
+)
+from eth_account.hdaccount import (
+    ETHEREUM_DEFAULT_PATH,
+)
+
+Account.enable_unaudited_hdwallet_features()
+
+
+@pytest.mark.parametrize("mnemonic,account_path,expected_address", [
+    # Ganache
+    # https://github.com/trufflesuite/ganache-core/blob/master/test/accounts.js
+    (
+        "into trim cross then helmet popular suit hammer cart shrug oval student",
+        ETHEREUM_DEFAULT_PATH,
+        "0x604a95C9165Bc95aE016a5299dd7d400dDDBEa9A",
+    ),
+    # Metamask
+    # https://github.com/MetaMask/eth-hd-keyring/blob/master/test/index.js
+    (
+        "finish oppose decorate face calm tragic certain desk hour urge dinosaur mango",
+        ETHEREUM_DEFAULT_PATH,
+        "0x1c96099350f13D558464eC79B9bE4445AA0eF579",
+    ),
+    (
+        "finish oppose decorate face calm tragic certain desk hour urge dinosaur mango",
+        "m/44'/60'/0'/0/1",  # 2nd account index in path
+        "0x1b00AeD43a693F3a957F9FeB5cC08AFA031E37a0",
+    ),
```

Comment:
Can we add a few tests to `from_mnemonic` please?
- account path with hardened indices (maybe one for a `'`, and one with an `H`, and one with `2**31 + i`?)
- account path more than 6 levels deep
- wrong passphrase

Reply:
So, "wrong passphrase" won't actually fail in any way, it will simply produce a different master seed than you expect.

I can do a check that creating an account with a passphrase and restoring with a different one will fail, if that sounds better

---

BIP44 Specifies this particular structure (6 levels deep). Practically it will work just the same for any depth, there are just no test vectors available for anything other than 6 levels deep (as no one uses it that way)

---

I can do a test case for a large value of account index, but still BIP43/BIP44 does not specify a hard node outside of those particular paths, so there won't be any test vectors available (outside of the ones in the HD test cases).

What I can do instead is add HDPath to the fuzzer? This would ensure that there is some alignment between how that feature is being implemented.

Reply:
> I can do a check that creating an account with a passphrase and restoring with a different one will fail, if that sounds better

Yep, that's what I meant!

> there are just no test vectors available for anything other than 6 levels deep (as no one uses it that way)

Ah interesting. I would have thought that people use more than the 6 levels quite frequently, but shows what I know ha.

> I can do a test case for a large value of account index, but still BIP43/BIP44 does not specify a hard node outside of those particular paths, so there won't be any test vectors available (outside of the ones in the HD test cases).

I'm not sure what you mean that BIP43/44 does not specify a hard node outside of those particular paths? It seems like you have all the official test vectors pretty well covered. At this point, I'm more concerned with the code UI working as we'd expect.

> What I can do instead is add HDPath to the fuzzer? This would ensure that there is some alignment between how that feature is being implemented.

Sounds good to me!

Reply:
> I would have thought that people use more than the 6 levels quite frequently

I think that would've been true when BIP32 was originally designed, but most people use the BIP43/BIP44 updates to BIP32. In BIP44, it defines a particular structure to how the path levels should be managed for purposes, coin types, change addresses, etc. Only certain nodes are hard nodes, because it's meant for a particular use case (like hardware wallets) where deriving the public address should not be possible. For reference:
https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#path-levels

Reply:
Yeah, I did read the BIP. I just figured that the account index would serve as the jumping off point for a user to make as many accounts as they wanted under that index, almost like a directory structure on a computer, rather than having them all on one level. But I guess if you have 2**31 - 1 addresses, no need to go deeper than that!

Reply:
HDPath Fuzzing added in df32c3b